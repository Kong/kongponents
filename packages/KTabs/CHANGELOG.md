# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.1.2](https://github.com/Kong/kongponents/compare/@kongponents/ktabs@0.1.1...@kongponents/ktabs@0.1.2) (2020-03-25)


### Bug Fixes

* ktabs add public ([#203](https://github.com/Kong/kongponents/issues/203)) ([8f5584b](https://github.com/Kong/kongponents/commit/8f5584bfef5c846d1154ab2ccb097e8f77189c13))





## 0.1.1 (2020-03-25)

**Note:** Version bump only for package @kongponents/ktabs
