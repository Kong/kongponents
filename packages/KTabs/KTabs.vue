<template>
  <div class="k-tabs">
    <ul
      role="tablist"
      aria-label="ktabs">
      <li
        v-for="(tab, i) in tabs"
        :key="tab.hash"
        :aria-selected="activeTab === tab.hash ? 'true' : 'false'"
        :aria-controls="`panel-${i}`"
        :id="`${tab.hash.replace('#','')}-tab`"
        :class="{ active: activeTab === tab.hash }"
        tabindex="0"
        role="tab"
        class="tab-item"
        @keydown.enter.prevent="activeTab = tab.hash"
        @click="handleTabChange(tab.hash)">
        <a class="tab-link">
          <slot :name="`${tab.hash.replace('#','')}-anchor`">{{ tab.title }}</slot>
        </a>
      </li>
    </ul>

    <div
      v-for="(tab, i) in tabs"
      :key="tab.hash"
      :id="`panel-${i}`"
      :aria-labelledby="`${tab.hash.replace('#','')}-tab`"
      role="tabpanel"
      tabindex="0"
      class="tab-container">
      <slot
        v-if="activeTab === tab.hash"
        :name="tab.hash.replace('#','')" />
    </div>
  </div>
</template>

<script>
export default {
  name: 'KTabs',
  // v-model
  model: {
    prop: 'value',
    event: 'input'
  },
  props: {
    /**
     * Array of Tab objects [{hash: '#tab1', title: 'tab1'}, {hash: '#tab2', title: 'tab2'}]
     */
    tabs: {
      type: Array,
      required: true
    },
    /**
     * A set tab hash to use as default
     */
    value: {
      type: String,
      default: ''
    }
  },

  data () {
    return {
      activeTab: this.value
        ? this.value
        : this.tabs[0].hash
    }
  },

  methods: {
    handleTabChange (tab) {
      this.activeTab = tab
      this.$emit('changed', tab)
    }
  }
}
</script>

<style lang="scss" scoped>
@import '~@kongponents/styles/_variables.scss';

.k-tabs {
  ul {
    display: flex;
    margin-bottom: var(--spacing-lg, spacing(lg));
    padding-left: 0;
    list-style: none;
    border-bottom: 2px solid var(--KTabsBottomBorder, var(--grey-88, color(grey-88)));
    .tab-item {
      position: relative;
      padding: var(--spacing-sm, spacing(sm));
      cursor: pointer;
      &:not(:first-of-type) { margin-left: var(--spacing-xs, spacing(xs)); }
      &:not(:last-of-type) { margin-right: var(--spacing-xs, spacing(xs)); }
      &:after {
        position: absolute;
        bottom: -2px;
        left: 0;
        width: 100%;
        height: 2px;
        display: block;
        content: '';
      }
      &.active,
      &:hover {
        &:after { background-color: var(--KTabsActiveColor, var(--blue-base, color(blue-base))); }
        .tab-link { color: var(--KTabsActiveColor, var(--blue-base, color(blue-base))); }
      }
    }
    .tab-link {
      display: block;
      color: var(--KTabsColor, var(--tblack-45, color(tblack-45)));
      &:hover {
        text-decoration: none;
        border: none;
      }
    }
  }
}
</style>
