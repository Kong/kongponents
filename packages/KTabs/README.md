# @kongponents/ktabs

[![](https://img.shields.io/npm/v/@kongponents/ktabs.svg?style=flat-square)](https://www.npmjs.com/package/@kongponents/ktabs)

```vue
<template>
  <KTabs :tabs="tabs">
    <template v-slot:tab1>
      <span>Tab 1 content</span>
    </template>
    <template v-slot:tab2>
      <span>Tab 2 content</span>
    </template>
  </KTabs>
</template>

<script>
export default {
  data() {
    return {
      tabs: [
        {
          hash: '#tab1', 
          title: 'tab1'
        },
        {
          hash: '#tab2',
          title: 'tab2
        }
      ]
    }
  }
}
</script>
```
