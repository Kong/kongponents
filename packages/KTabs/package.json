{
  "name": "@kongponents/ktabs",
  "version": "5.0.7",
  "description": "Tabs component.",
  "main": "dist/KTabs.umd.min.js",
  "componentName": "KTabs",
  "source": "KTabs.vue",
  "files": [
    "dist",
    "KTabs.vue"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Kong/kongponents.git"
  },
  "author": "Kong Inc",
  "license": "Apache-2.0",
  "bugs": {
    "url": "https://github.com/Kong/kongponents/issues"
  },
  "homepage": "https://github.com/Kong/kongponents#readme",
  "publishConfig": {
    "access": "public"
  }
}
