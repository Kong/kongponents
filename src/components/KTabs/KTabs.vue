<template>
  <div class="k-tabs">
    <ul
      aria-label="Tabs"
      role="tablist"
    >
      <li
        v-for="tab in tabs"
        :id="`${tab.hash.replace('#','')}-tab`"
        :key="tab.hash"
        class="tab-item"
        :class="{ active: activeTab === tab.hash }"
      >
        <div
          :aria-controls="hidePanels ? undefined : `panel-${tab.hash}`"
          :aria-selected="hidePanels ? undefined : (activeTab === tab.hash ? 'true' : 'false')"
          class="tab-link"
          :class="{ 'has-panels': !hidePanels, disabled: tab.disabled }"
          role="tab"
          :tabindex="tab.disabled ? '-1' : '0'"
          @click.prevent="!tab.disabled ? handleTabChange(tab.hash) : undefined"
          @keydown.enter.prevent="!tab.disabled ? handleTabChange(tab.hash) : undefined"
          @keydown.space.prevent="!tab.disabled ? handleTabChange(tab.hash) : undefined"
        >
          <slot :name="`${tab.hash.replace('#','')}-anchor`">
            <span>{{ tab.title }}</span>
          </slot>
        </div>
      </li>
    </ul>

    <template v-if="!hidePanels">
      <div
        v-for="(tab, i) in tabs"
        :id="`panel-${i}`"
        :key="tab.hash"
        :aria-labelledby="`${tab.hash.replace('#','')}-tab`"
        class="tab-container"
        role="tabpanel"
      >
        <slot
          v-if="activeTab === tab.hash"
          :name="tab.hash.replace('#','')"
        />
      </div>
    </template>
  </div>
</template>

<script setup lang="ts">
import type { PropType } from 'vue'
import { ref, watch } from 'vue'
import type { Tab } from '@/types'

const props = defineProps({
  /**
   * Array of Tab objects [{ hash: '#tab1', title: 'tab1'}, {hash: '#tab2', title: 'tab2' }]
   */
  tabs: {
    type: Array as PropType<Tab[]>,
    required: true,
  },
  /**
   * A set tab hash to use as default
   */
  modelValue: {
    type: String,
    default: '',
    validator: (val: string): boolean => val === '' || (val.includes('#') && !val.includes(' ')),
  },
  /**
   * Render the tab's corresponding panel container
   */
  hidePanels: {
    type: Boolean,
    default: false,
  },
})

const emit = defineEmits<{
  (e: 'update:modelValue', val: string): void
  (e: 'change', val: string): void
}>()

const activeTab = ref<string>(props.modelValue ? props.modelValue : props.tabs[0].hash)

const handleTabChange = (tab: string): void => {
  activeTab.value = tab
  emit('change', tab)
  emit('update:modelValue', tab)
}

watch(() => props.modelValue, (newTabHash) => {
  activeTab.value = newTabHash
  emit('change', newTabHash)
  emit('update:modelValue', newTabHash)
})
</script>

<style lang="scss" scoped>
.k-tabs {
  ul {
    border-bottom: var(--kui-border-width-10, $kui-border-width-10) solid var(--kui-color-border, $kui-color-border);
    display: flex;
    font-family: var(--kui-font-family-text, $kui-font-family-text);
    font-size: var(--kui-font-size-30, $kui-font-size-30);
    font-weight: var(--kui-font-weight-semibold, $kui-font-weight-semibold);
    gap: var(--kui-space-40, $kui-space-40);
    line-height: var(--kui-line-height-40, $kui-line-height-40);
    list-style: none;
    margin-bottom: var(--kui-space-70, $kui-space-70);
    margin-top: var(--kui-space-0, $kui-space-0);
    overflow-x: auto;
    overflow-y: hidden;
    padding: var(--kui-space-0, $kui-space-0) var(--kui-space-70, $kui-space-70);
    padding-top: var(--kui-space-20, $kui-space-20);

    .tab-item {
      border-bottom-color: var(--kui-color-border-transparent, $kui-color-border-transparent);
      border-bottom-style: solid;
      border-bottom-width: var(--kui-border-width-20, $kui-border-width-20);
      padding-bottom: var(--kui-space-40, $kui-space-40);
      position: relative;
      transition: border-bottom-color $kongponentsTransitionDurTimingFunc;
      white-space: nowrap;

      .tab-link {
        border-radius: var(--kui-border-radius-30, $kui-border-radius-30);
        color: var(--kui-color-text-neutral, $kui-color-text-neutral);
        cursor: pointer;
        display: inline-flex;
        gap: var(--kui-space-40, $kui-space-40);
        text-decoration: none;
        transition: color $kongponentsTransitionDurTimingFunc, background-color $kongponentsTransitionDurTimingFunc, box-shadow $kongponentsTransitionDurTimingFunc;
        user-select: none;

        // Applies the padding to the tab’s content when not showing panels which is typically used for placing links inside KTabs for navigational tabs. Otherwise, clicking the tab outside of the link’s box will mark it as active but won’t actually navigate.
        &.has-panels,
        &:not(.has-panels) :deep(> *) {
          padding: var(--kui-space-30, $kui-space-30) var(--kui-space-50, $kui-space-50);
        }

        a, :deep(a) {
          color: var(--kui-color-text, $kui-color-text);
          text-decoration: none;
        }

        &:hover:not(.disabled) {
          background-color: var(--kui-color-background-neutral-weaker, $kui-color-background-neutral-weaker);
        }

        &:focus-visible {
          background-color: var(--kui-color-background-neutral-weaker, $kui-color-background-neutral-weaker);
          box-shadow: var(--kui-shadow-focus, $kui-shadow-focus);
          color: var(--kui-color-text, $kui-color-text);
          outline: none;
        }

        &.disabled {
          color: var(--kui-color-text-disabled, $kui-color-text-disabled);
          cursor: not-allowed;
        }
      }

      &.active {
        border-bottom-color: var(--kui-color-border-decorative-purple, $kui-color-border-decorative-purple);

        .tab-link {
          color: var(--kui-color-text, $kui-color-text);
        }
      }
    }
  }
}
</style>
