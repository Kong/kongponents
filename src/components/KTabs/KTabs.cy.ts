import { mount } from 'cypress/vue'
import { h } from 'vue'
import KTabs from '@/components/KTabs/KTabs.vue'

const TABS = [
  { hash: '#pictures', title: 'Pictures' },
  { hash: '#movies', title: 'Movies' },
  { hash: '#books', title: 'Books' },
]

describe('KTabs', () => {
  it('first tab is set if hash not found', () => {
    mount(KTabs, {
      props: {
        tabs: TABS,
      },
    })

    cy.get('.tab-item').eq(0).should('have.class', 'active')
  })

  it('sets correct tab if default tab prop', () => {
    mount(KTabs, {
      props: {
        tabs: TABS,
        modelValue: '#books',
      },
    })

    cy.get('.tab-item').eq(2).should('have.class', 'active')
  })

  it('emits change event on click', () => {
    mount(KTabs, {
      props: {
        tabs: TABS,
      },
    })

    cy.get('.tab-item').eq(1).click().then(() => {
      cy.wrap(Cypress.vueWrapper.emitted()).should('have.property', 'change')
      cy.wrap(Cypress.vueWrapper.emitted('change')[0][0]).should('eq', '#movies')
    })
  })

  it('hides the panel content when `hidePanels` is true', () => {
    const picturesSlot = 'I love pictures'
    const moviesSlot = 'I love pictures'
    const booksSlot = 'I love pictures'

    mount(KTabs, {
      props: {
        tabs: TABS,
        hidePanels: true,
      },
      slots: {
        pictures: h('div', {}, picturesSlot),
        movies: h('div', {}, moviesSlot),
        books: h('div', {}, booksSlot),
      },
    })

    cy.get('.tab-item').eq(0).click().then(() => {
      cy.get('#panel-0').should('not.exist')
      cy.get('.tab-container').should('not.exist')
    })
    cy.get('.tab-item').eq(1).click().then(() => {
      cy.get('#panel-1').should('not.exist')
      cy.get('.tab-container').should('not.exist')
    })
    cy.get('.tab-item').eq(2).click().then(() => {
      cy.get('#panel-2').should('not.exist')
      cy.get('.tab-container').should('not.exist')
    })
  })

  describe('slots', () => {
    it('provides the #hash slot content', () => {
      const picturesSlot = 'I love pictures'
      const moviesSlot = 'I love pictures'
      const booksSlot = 'I love pictures'

      mount(KTabs, {
        props: {
          tabs: TABS,
        },
        slots: {
          pictures: h('div', {}, picturesSlot),
          movies: h('div', {}, moviesSlot),
          books: h('div', {}, booksSlot),
        },
      })

      cy.get('.tab-item').eq(0).click().then(() => {
        cy.get('#panel-0').should('contain.text', picturesSlot)
      })
      cy.get('.tab-item').eq(1).click().then(() => {
        cy.get('#panel-1').should('contain.text', moviesSlot)
      })
      cy.get('.tab-item').eq(2).click().then(() => {
        cy.get('#panel-2').should('contain.text', booksSlot)
      })
    })

    it('provides the anchor slot content', () => {
      const picturesSlot = 'I love pictures'
      const moviesSlot = 'I love pictures'
      const booksSlot = 'I love pictures'

      mount(KTabs, {
        props: {
          tabs: TABS,
        },
        slots: {
          'pictures-anchor': h('div', {}, picturesSlot),
          'movies-anchor': h('div', {}, moviesSlot),
          'books-anchor': h('div', {}, booksSlot),
        },
      })

      cy.get('#pictures-tab .tab-link').should('contain.text', picturesSlot)
      cy.get('#movies-tab .tab-link').should('contain.text', moviesSlot)
      cy.get('#books-tab .tab-link').should('contain.text', booksSlot)
    })
  })
})
