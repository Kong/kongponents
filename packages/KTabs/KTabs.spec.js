import { mount } from '@vue/test-utils'
import KTabs from '@/KTabs/KTabs'

describe('KTabs', () => {
  it('matches snapshot', () => {
    const wrapper = mount(KTabs)

    expect(wrapper.html()).toMatchSnapshot()
  })
})
