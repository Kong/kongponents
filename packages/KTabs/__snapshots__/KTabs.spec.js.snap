// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KTabs matches snapshot 1`] = `
<div class="k-tabs">
  <ul role="tablist" aria-label="ktabs">
    <li aria-selected="true" aria-controls="panel-0" id="pictures-tab" tabindex="0" role="tab" class="tab-item active"><a class="tab-link">Pictures</a></li>
    <li aria-selected="false" aria-controls="panel-1" id="movies-tab" tabindex="0" role="tab" class="tab-item"><a class="tab-link">Movies</a></li>
    <li aria-selected="false" aria-controls="panel-2" id="books-tab" tabindex="0" role="tab" class="tab-item"><a class="tab-link">Books</a></li>
  </ul>
  <div id="panel-0" aria-labelledby="pictures-tab" role="tabpanel" tabindex="0" class="tab-container"></div>
  <div id="panel-1" aria-labelledby="movies-tab" role="tabpanel" tabindex="0" class="tab-container">
    <!---->
  </div>
  <div id="panel-2" aria-labelledby="books-tab" role="tabpanel" tabindex="0" class="tab-container">
    <!---->
  </div>
</div>
`;
